src/UhdmAst.cpp

```diff
@@ -607,40 +606,41 @@ AstNodeDType* getDType(FileLine* fl, vpiHandle obj_h, UhdmShared& shared) {
         break;
     }
     case vpiLogicTypespec:
-    case vpiBitTypespec: {
-        if (auto elem_typespec_h = vpi_handle(vpiElemTypespec, obj_h)) {
-            dtypep = VN_CAST(process_typespec(elem_typespec_h, shared), NodeDType);
-            if (!dtypep)
-                v3error("Unable to handle vpiElemTypespec node as AstNodeDType");
-        } else {
-            AstBasicDTypeKwd keyword = get_kwd_for_type(type);
-            dtypep = new AstBasicDType(fl, keyword);
-        }
-        dtypep = applyPackedRanges(fl, obj_h, dtypep, shared);
-        break;
-    }
-    case vpiArrayTypespec: {
-        auto elem_typespec_h = vpi_handle(vpiElemTypespec, obj_h);
-        dtypep = getDType(fl, elem_typespec_h, shared);
-        dtypep = applyUnpackedRanges(fl, obj_h, dtypep, shared);
-        break;
-    }
-    case vpiPackedArrayTypespec: {
-        auto elem_typespec_h = vpi_handle(vpiElemTypespec, obj_h);
-        if (elem_typespec_h) {
-            dtypep = getDType(fl, elem_typespec_h, shared);
+    case vpiBitTypespec:
+    case vpiIntTypespec:
+    case vpiLongIntTypespec:
+    case vpiIntegerTypespec:
+    case vpiByteTypespec:
+    case vpiRealTypespec:
+    case vpiStringTypespec:
+    case vpiChandleTypespec:
+    case vpiTimeTypespec:
+    case vpiArrayTypespec:
+    case vpiPackedArrayTypespec:
+    case vpiStructTypespec:
+    case vpiEnumTypespec:
+    case vpiUnionTypespec: {
+        std::string typespec_name = get_object_name(obj_h);
+        std::string full_type_name;
+        if (typespec_name != "") {
+            auto pos = typespec_name.rfind("::");
+            if (pos != std::string::npos) typespec_name = typespec_name.substr(pos + 2);
+
+            std::string package_name = "";
+            if (vpiHandle instance_h = vpi_handle(vpiInstance, obj_h))
+                package_name = get_object_name(instance_h, {vpiDefName});
+            if (package_name != "")
+                full_type_name = package_name + "::" + typespec_name;
+            else
+                full_type_name = typespec_name;
+            dtypep = get_type_reference(fl, typespec_name, full_type_name, shared);
+            break;
         } else {
-            UINFO(7, "No elem_typespec found in vpiPackedArrayTypespec, trying IndexTypespec"
-                         << std::endl);
-            auto index_typespec_h = vpi_handle(vpiIndexTypespec, obj_h);
-            if (index_typespec_h) {
-                dtypep = getDType(fl, index_typespec_h, shared);
-            } else {
-                v3error("\t! Failed to get typespec handle for PackedArrayTypespec");
-            }
+            // Typespec without name, construct the type by process_typespec
+            dtypep = VN_CAST(process_typespec(obj_h, shared), NodeDType);
+            if (!dtypep) v3error("Unable to handle vpiTypespec node");
```

Comment:
This gives the exact same error as [line 600](https://github.com/antmicro/verilator/pull/413/files#diff-eab5a9ad66f2411094af89dec2064f768624fd994a7a51ea6244c67cfc794c15R600), please try to provide more context to aid debugging.